sample_consensus/include/pcl/sample_consensus/impl/ransac.hpp

```diff
@@ -55,77 +58,148 @@ pcl::RandomSampleConsensus<PointT>::computeModel (int)
+  {
+#if defined _OPENMP && _OPENMP >= 201107 // We need OpenMP 3.1 for the atomic constructs
+    if (threads_ == 0)
     {
-      PCL_DEBUG ("[pcl::RandomSampleConsensus::computeModel] RANSAC reached the maximum number of trials.\n");
-      break;
+      threads_ = omp_get_num_procs();
+      PCL_DEBUG ("[pcl::RandomSampleConsensus::computeModel] Automatic number of threads requested, choosing %i threads.\n", threads_);
     }
+#else
+    // Parallelization desired, but not available
+    PCL_WARN ("[pcl::RandomSampleConsensus::computeModel] Parallelization is requested, but OpenMP 3.1 is not available! Continuing without parallelization.\n");
+    threads_ = -1;
+#endif
   }
 
-  PCL_DEBUG ("[pcl::RandomSampleConsensus::computeModel] Model: %lu size, %d inliers.\n", model_.size (), n_best_inliers_count);
+#if defined _OPENMP && _OPENMP >= 201107 // We need OpenMP 3.1 for the atomic constructs
+#pragma omp parallel if(threads_ > 0) num_threads(threads_) shared(k, skipped_count, n_best_inliers_count) private(selection, model_coefficients, n_inliers_count) // would be nice to have a default(none)-clause here, but then some compilers complain about the shared const variables
+#endif
+  {
+#if defined _OPENMP && _OPENMP >= 201107
+    if (omp_in_parallel())
+#pragma omp master
+      PCL_DEBUG ("[pcl::RandomSampleConsensus::computeModel] Computing in parallel with up to %i threads.\n", omp_get_num_threads());
+    else
+#endif
+      PCL_DEBUG ("[pcl::RandomSampleConsensus::computeModel] Computing not parallel.\n");
+
+    // Iterate
+    while (true) // infinite loop with four possible breaks
+    {
+      // Get X samples which satisfy the model criteria
+#if defined _OPENMP && _OPENMP >= 201107
+#pragma omp critical(samples)
+#endif
+      {
+        sac_model_->getSamples (iterations_, selection); // The random number generator used when choosing the samples should not be called in parallel
+      }
+
+      if (selection.empty ())
+      {
+        PCL_ERROR ("[pcl::RandomSampleConsensus::computeModel] No samples could be selected!\n");
+        break;
+      }
+
+      // Search for inliers in the point cloud for the current plane model M
+      if (!sac_model_->computeModelCoefficients (selection, model_coefficients)) // This function has to be thread-safe
+      {
+        //++iterations_;
+        unsigned skipped_count_tmp;
+#if defined _OPENMP && _OPENMP >= 201107
+#pragma omp atomic capture
+#endif
+        skipped_count_tmp = ++skipped_count;
+        if (skipped_count_tmp < max_skip)
+          continue;
+        else
+          break;
+      }
+
+      // Select the inliers that are within threshold_ from the model
+      //sac_model_->selectWithinDistance (model_coefficients, threshold_, inliers);
+      //if (inliers.empty () && k > 1.0)
+      //  continue;
+
+      n_inliers_count = sac_model_->countWithinDistance (model_coefficients, threshold_); // This functions has to be thread-safe. Most work is done here
```

Comment:
Could you please copy this comment to someplace in a doc comment? Seems important 😄

Reply:
Since both `computeModelCoefficients` and `countWithinDistance` are [declared const](https://github.com/PointCloudLibrary/pcl/blob/master/sample_consensus/include/pcl/sample_consensus/sac_model.h#L259), they cannot change any member variables, which makes them thread-safe. Or am I mistaken? (I can still add that to the funciton description)

Reply:
Having assumptions in one place can prevent future mistakes. Merely marking the functions  in the correct place (the one you linked) can help the person refactoring/developing a SAC Model